language/move-lang/src/expansion/dependency_ordering.rs

```diff
@@ -46,44 +62,79 @@ enum DepType {
+    // A summary of neighbors keyed by module or script
+    neighbors_by_unit: BTreeMap<ModuleOrScriptIdent, BTreeSet<E::Neighbor>>,
+    // The module or script we are currently exploring
+    current: Option<ModuleOrScriptIdent>,
 }
 
 impl<'a> Context<'a> {
     fn new(modules: &'a UniqueMap<ModuleIdent, E::ModuleDefinition>) -> Self {
         Context {
             modules,
-            neighbors: BTreeMap::new(),
-            current_module: None,
+            module_neighbors: BTreeMap::new(),
+            neighbors_by_unit: BTreeMap::new(),
+            current: None,
         }
     }
 
     fn add_neighbor(&mut self, mident: ModuleIdent, dep_type: DepType, loc: Loc) {
-        let current_mident = self.current_module.clone().unwrap();
-        if current_mident == mident || !self.modules.contains_key(&mident) {
+        if !self.modules.contains_key(&mident) {
+            // as the dependency checking happens before the naming phase, it is possible to refer
+            // to a module with a ModuleIdent outside of the compilation context. Do not add such
+            // modules as neighbors.
             return;
         }
-        let (node, new_neighbor) = match dep_type {
-            DepType::Use => (current_mident, mident),
-            DepType::Friend => (mident, current_mident),
-        };
 
-        let m = self
-            .neighbors
-            .entry(node)
-            .or_insert_with(BTreeMap::new)
-            .entry(new_neighbor)
-            .or_insert_with(BTreeMap::new);
-        if m.contains_key(&dep_type) {
+        let current = self.current.clone().unwrap();
+        if matches!(&current, ModuleOrScriptIdent::Module(current_mident) if &mident == current_mident)
+        {
+            // do not add the module itself as a neighbor
             return;
         }
-        m.insert(dep_type, loc);
+
+        let neighbor = match dep_type {
+            DepType::Use => E::Neighbor::Dependency(mident.clone()),
+            DepType::Friend => E::Neighbor::Friend(mident.clone()),
+        };
+        self.neighbors_by_unit
+            .entry(current.clone())
+            .or_insert_with(BTreeSet::new)
+            .insert(neighbor);
+
+        match current {
+            ModuleOrScriptIdent::Module(current_mident) => {
+                let (node, new_neighbor) = match dep_type {
+                    DepType::Use => (current_mident, mident),
+                    DepType::Friend => (mident, current_mident),
+                };
+                let m = self
+                    .module_neighbors
+                    .entry(node)
+                    .or_insert_with(BTreeMap::new)
+                    .entry(new_neighbor)
+                    .or_insert_with(BTreeMap::new);
+                if m.contains_key(&dep_type) {
+                    return;
+                }
+                m.insert(dep_type, loc);
+            }
+            ModuleOrScriptIdent::Script(..) => {}
```

Comment:
nit `Script(_)` as a general style in the compiler I use underscores when I can for positional fields and only use `..` for named fields. Can make migrations a bit easier

Reply:
Also, nit (), instead of {}

Reply:
Fixed.